<a href="https://colab.research.google.com/github/Jeetanand/HR-Analytics-Employee-Attrition/blob/main/HR_attrition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
hrattr_data = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

In [4]:
hrattr_data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [6]:
# converting Yes or No categories into 1 and 0
hrattr_data['Attrition_ind'] = 0
hrattr_data.loc[hrattr_data['Attrition'] =='Yes', 'Attrition_ind'] = 1

In [7]:
dummy_busnstrvl = pd.get_dummies(hrattr_data['BusinessTravel'], prefix='busns_trvl')

In [8]:
dummy_busnstrvl

busns_trvl_Non-Travel  busns_trvl_Travel_Frequently  \
0                         0                             0   
1                         0                             1   
2                         0                             0   
3                         0                             1   
4                         0                             0   
...                     ...                           ...   
1465                      0                             1   
1466                      0                             0   
1467                      0                             0   
1468                      0                             1   
1469                      0                             0   

      busns_trvl_Travel_Rarely  
0                            1  
1                            0  
2                            1  
3                            0  
4                            1  
...                        ...  
1465                         0  
1466                         1  
1467                         1  
1468                         0  
1469                         1  

[1470 rows x 3 columns]

In [9]:
dummy_dept = pd.get_dummies(hrattr_data['Department'], prefix='dept')

In [10]:
dummy_dept

dept_Human Resources  dept_Research & Development  dept_Sales
0                        0                            0           1
1                        0                            1           0
2                        0                            1           0
3                        0                            1           0
4                        0                            1           0
...                    ...                          ...         ...
1465                     0                            1           0
1466                     0                            1           0
1467                     0                            1           0
1468                     0                            0           1
1469                     0                            1           0

[1470 rows x 3 columns]

In [11]:
dummy_edufield = pd.get_dummies(hrattr_data['EducationField'], prefix='edufield')

In [12]:
dummy_edufield

edufield_Human Resources  edufield_Life Sciences  edufield_Marketing  \
0                            0                       1                   0   
1                            0                       1                   0   
2                            0                       0                   0   
3                            0                       1                   0   
4                            0                       0                   0   
...                        ...                     ...                 ...   
1465                         0                       0                   0   
1466                         0                       0                   0   
1467                         0                       1                   0   
1468                         0                       0                   0   
1469                         0                       0                   0   

      edufield_Medical  edufield_Other  edufield_Technical Degree  
0                    0               0                          0  
1                    0               0                          0  
2                    0               1                          0  
3                    0               0                          0  
4                    1               0                          0  
...                ...             ...                        ...  
1465                 1               0                          0  
1466                 1               0                          0  
1467                 0               0                          0  
1468                 1               0                          0  
1469                 1               0                          0  

[1470 rows x 6 columns]

In [13]:
dummy_gender = pd.get_dummies(hrattr_data['Gender'], prefix='gend')

In [14]:
dummy_gender

gend_Female  gend_Male
0               1          0
1               0          1
2               0          1
3               1          0
4               0          1
...           ...        ...
1465            0          1
1466            0          1
1467            0          1
1468            0          1
1469            0          1

[1470 rows x 2 columns]

In [15]:
dummy_jobrole = pd.get_dummies(hrattr_data['JobRole'], prefix='jobrole')

In [16]:
dummy_jobrole

jobrole_Healthcare Representative  jobrole_Human Resources  \
0                                     0                        0   
1                                     0                        0   
2                                     0                        0   
3                                     0                        0   
4                                     0                        0   
...                                 ...                      ...   
1465                                  0                        0   
1466                                  1                        0   
1467                                  0                        0   
1468                                  0                        0   
1469                                  0                        0   

      jobrole_Laboratory Technician  jobrole_Manager  \
0                                 0                0   
1                                 0                0   
2                                 1                0   
3                                 0                0   
4                                 1                0   
...                             ...              ...   
1465                              1                0   
1466                              0                0   
1467                              0                0   
1468                              0                0   
1469                              1                0   

      jobrole_Manufacturing Director  jobrole_Research Director  \
0                                  0                          0   
1                                  0                          0   
2                                  0                          0   
3                                  0                          0   
4                                  0                          0   
...                              ...                        ...   
1465                               0                          0   
1466                               0                          0   
1467                               1                          0   
1468                               0                          0   
1469                               0                          0   

      jobrole_Research Scientist  jobrole_Sales Executive  \
0                              0                        1   
1                              1                        0   
2                              0                        0   
3                              1                        0   
4                              0                        0   
...                          ...                      ...   
1465                           0                        0   
1466                           0                        0   
1467                           0                        0   
1468                           0                        1   
1469                           0                        0   

      jobrole_Sales Representative  
0                                0  
1                                0  
2                                0  
3                                0  
4                                0  
...                            ...  
1465                             0  
1466                             0  
1467                             0  
1468                             0  
1469                             0  

[1470 rows x 9 columns]

In [17]:
dummy_maritstat = pd.get_dummies(hrattr_data['MaritalStatus'], prefix='maritalstat')

In [18]:
dummy_maritstat

maritalstat_Divorced  maritalstat_Married  maritalstat_Single
0                        0                    0                   1
1                        0                    1                   0
2                        0                    0                   1
3                        0                    1                   0
4                        0                    1                   0
...                    ...                  ...                 ...
1465                     0                    1                   0
1466                     0                    1                   0
1467                     0                    1                   0
1468                     0                    1                   0
1469                     0                    1                   0

[1470 rows x 3 columns]

In [19]:
dummy_overtime = pd.get_dummies(hrattr_data['OverTime'], prefix='overtime')

In [20]:
dummy_overtime

overtime_No  overtime_Yes
0               0             1
1               1             0
2               0             1
3               0             1
4               1             0
...           ...           ...
1465            1             0
1466            1             0
1467            0             1
1468            1             0
1469            1             0

[1470 rows x 2 columns]

**Seperating Continous variable**


In [24]:
continuous_columns = ['Age','DailyRate','DistanceFromHome','Education', 'EnvironmentSatisfaction','HourlyRate','JobInvolvement','JobLevel','JobSatisfaction', 'MonthlyIncome', 'MonthlyRate','NumCompaniesWorked','PercentSalaryHike', 'PerformanceRating','RelationshipSatisfaction','StockOptionLevel', 'TotalWorkingYears','TrainingTimesLastYear','WorkLifeBalance', 'YearsAtCompany','YearsInCurrentRole', 'YearsSinceLastPromotion','YearsWithCurrManager']

In [25]:
continuous_columns

['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [26]:
hrattr_continuous = hrattr_data[continuous_columns]

In [27]:
hrattr_data_new = pd.concat([dummy_busnstrvl, dummy_dept,
dummy_edufield, dummy_gender, dummy_jobrole, dummy_maritstat,
dummy_overtime, hrattr_continuous, hrattr_data['Attrition_ind']],axis=1)

In [29]:
# Train and Test split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split( hrattr_data_new.drop
(['Attrition_ind'], axis=1),hrattr_data_new['Attrition_ind'], train_size
= 0.7, random_state=42)

**Decision tree classifier**

In [30]:
# Decision Tree Classifier
>>> from sklearn.tree import DecisionTreeClassifier

The parameters selected for the DT classifier are in the following code with splitting
criterion as Gini, Maximum depth as 5, minimum number of observations required for
qualifying split is 2, and the minimum samples that should be present in the terminal node
is 1:

In [32]:
dt_fit = DecisionTreeClassifier(criterion="gini",max_depth=5,min_samples_split=2, min_samples_leaf=1,random_state=42)
dt_fit.fit(x_train,y_train)

DecisionTreeClassifier(max_depth=5, random_state=42)

In [33]:
print ("\nDecision Tree - Train Confusion Matrix\n\n", pd.crosstab(y_train, dt_fit.predict(x_train),rownames =["Actuall"],colnames = ["Predicted"]))


Decision Tree - Train Confusion Matrix

 Predicted    0   1
Actuall           
0          844   9
1           98  78


In [34]:
from sklearn.metrics import accuracy_score, classification_report

In [35]:
print ("\nDecision Tree - Train accuracy\n\n",round(accuracy_score(y_train, dt_fit.predict(x_train)),3))


Decision Tree - Train accuracy

 0.896


In [36]:
print ("\nDecision Tree - Train Classification Report\n", classification_report(y_train, dt_fit.predict(x_train)))


Decision Tree - Train Classification Report
               precision    recall  f1-score   support

           0       0.90      0.99      0.94       853
           1       0.90      0.44      0.59       176

    accuracy                           0.90      1029
   macro avg       0.90      0.72      0.77      1029
weighted avg       0.90      0.90      0.88      1029



In [37]:
print ("\n\nDecision Tree - Test ConfusionMatrix\n\n",pd.crosstab(y_test, dt_fit.predict(x_test),rownames =["Actuall"],colnames = ["Predicted"]))



Decision Tree - Test ConfusionMatrix

 Predicted    0   1
Actuall           
0          361  19
1           49  12


In [38]:
print ("\nDecision Tree - Test accuracy",round(accuracy_score(y_test, dt_fit.predict(x_test)),3))


Decision Tree - Test accuracy 0.846


In [39]:
print ("\nDecision Tree - Test Classification Report\n", classification_report( y_test, dt_fit.predict(x_test)))


Decision Tree - Test Classification Report
               precision    recall  f1-score   support

           0       0.88      0.95      0.91       380
           1       0.39      0.20      0.26        61

    accuracy                           0.85       441
   macro avg       0.63      0.57      0.59       441
weighted avg       0.81      0.85      0.82       441



By carefully observing the results, we can infer that, even though the test accuracy is high
(84.6%), the precision and recall of one category (Attrition = Yes) is low (precision = 0.39 and
recall = 0.20). This could be a serious issue when management tries to use this model to
provide some extra benefits proactively to the employees with a high chance of attrition
prior to actual attrition, as this model is unable to identify the real employees who will be
leaving.

**Tuning class weights in decision tree classifier**

In [49]:
import numpy as np
dummyarray = np.empty((6,10))
dt_wttune = pd.DataFrame(dummyarray)

In [50]:
dt_wttune.columns = ["zero_wght","one_wght","tr_accuracy",
"tst_accuracy", "prec_zero","prec_one", "prec_ovll",
"recl_zero","recl_one","recl_ovll"]

In [51]:
zero_clwghts = [0.01,0.1,0.2,0.3,0.4,0.5]

In [54]:
for i in range(len(zero_clwghts)):
  clwght = {0:zero_clwghts[i],1:1.0-zero_clwghts[i]}
  dt_fit = DecisionTreeClassifier(criterion="gini", max_depth=5, min_samples_split=2, min_samples_leaf=1,random_state=42,class_weight =clwght)
  dt_fit.fit(x_train,y_train)
  dt_wttune.loc[i, 'zero_wght'] = clwght[0]
  dt_wttune.loc[i, 'one_wght'] = clwght[1]
  dt_wttune.loc[i, 'tr_accuracy'] = round(accuracy_score(y_train, dt_fit.predict( x_train)),3)
  dt_wttune.loc[i, 'tst_accuracy'] = round(accuracy_score(y_test,dt_fit.predict( x_test)),3)
  clf_sp = classification_report(y_test,dt_fit.predict(x_test)).split()
  dt_wttune.loc[i, 'prec_zero'] = float(clf_sp[5])
  dt_wttune.loc[i, 'prec_one'] = float(clf_sp[10])
  dt_wttune.loc[i, 'prec_ovll'] = float(clf_sp[17])
  dt_wttune.loc[i, 'recl_zero'] = float(clf_sp[6])
  dt_wttune.loc[i, 'recl_one'] = float(clf_sp[11])
  dt_wttune.loc[i, 'recl_ovll'] = float(clf_sp[18])
  print ("\nClass Weights",clwght,"Trainaccuracy:",round(accuracy_score( y_train,dt_fit.predict(x_train)),3),"Testaccuracy:",round(accuracy_score(y_test, dt_fit.predict(x_test)),3))
  print ("Test ConfusionMatrix\n\n",pd.crosstab(y_test,dt_fit.predict( x_test),rownames =["Actuall"],colnames = ["Predicted"]))

ValueError: ignored